LSP to resolve custom include paths (e.g. Hardhat)

christianparpart opened the pull request:
This PR is WIP, but it will fix #12833 later on.

Big question is, how do we want to tell `solc --lsp` where to look for include paths. I remember there was some project file format that does include that information, but what was that? @cameel, do you remember?

### Checklist:

- [x] factor our file path resolving in FileRepository: `tryResolvePath(.)`
- [x] fix `sourceUnitNameToUri()` by making use of new path resolver.
- [x] custom include paths via JSON settings object
- [x] always add `{projectRoot}/node_modules` to include paths (vscode-solidity is doing that too and users are used to it)

cameel commented:
> I remember there was some project file format that does include that information, but what was that? @cameel, do you remember?

Do you mean the dappfile ([EIP 82](https://github.com/ethereum/EIPs/issues/82))? Or one of the config files actually used by frameworks?

christianparpart commented:
> > I remember there was some project file format that does include that information, but what was that? @cameel, do you remember?
> 
> Do you mean the dappfile ([EIP 82](https://github.com/ethereum/EIPs/issues/82))? Or one of the config files actually used by frameworks?

ah right thanks! Dappfile is what I remembered, but this time, I in fact meant a framework that introduced a file that is now also supported by another one. I am talking vaguely because I forgot the names. ;-(

I might look into Dappfile though.

christianparpart committed:
lsp: Adds support for include paths and have {projectUri}/node_modules added by default.

- Factor out FileRepository's path resolving into own public function.
- Fixes sourceUnitNameToUri() path resolving in relation to include paths being used.
- Adding an solAssert().
- adds nother test for include-paths (bad include)
- Fixes a case on Windows there an ill-formed URI was generated.
- Dropping unnecessary if-branch when translating from sourceUnitName to URI.

cameel commented:
> ah right thanks! Dappfile is what I remembered, but this time, I in fact meant a framework that introduced a file that is now also supported by another one. I am talking vaguely because I forgot the names. ;-(

Oh, you meant `remappings.txt` (https://github.com/ethereum/solidity/issues/12603).

cameel commented:
> if we can't come up with something immediately, what do you think about adding a CLI option,

Why not just support `--include-paths` instead of adding a new arg?

christianparpart commented:
> > if we can't come up with something immediately, what do you think about adding a CLI option,
> 
> Why not just support `--include-paths` instead of adding a new arg?

I can do that. thanks. But I am still hoping for some kind of auto-configuration, where users do not need to do too much.

Suppose you are working in more than one project, and they all do not necessarily share the same include paths.

I think we could auto-detect at least the well known frameworks and adjust the include paths accordingly (on top of manually added ones, if added). What do you think, @cameel ?

stackenbotten commented on libsolidity/lsp/LanguageServer.cpp:
```diff
@@ -101,6 +101,21 @@ Json::Value LanguageServer::toJson(SourceLocation const& _location)
+	{
+		std::vector<boost::filesystem::path> includePaths;
+		Json::Value jsonIncludePaths = _settings["includes"];
+		for (Json::Value jsonPath : jsonIncludePaths)
```
Coding style error

stackenbotten commented on libsolidity/lsp/LanguageServer.cpp:
```diff
@@ -101,14 +101,29 @@ Json::Value LanguageServer::toJson(SourceLocation const& _location)
+	if (jsonIncludePaths && jsonIncludePaths.isArray())
+	{
+		std::vector<boost::filesystem::path> includePaths;
+		for (Json::Value jsonPath : jsonIncludePaths)
```
Coding style error

cameel commented on libsolidity/lsp/LanguageServer.cpp:
```diff
@@ -101,14 +101,29 @@ Json::Value LanguageServer::toJson(SourceLocation const& _location)
 {
 	m_settingsObject = _settings;
+	Json::Value jsonIncludePaths = _settings["include-paths"];
+	if (jsonIncludePaths && jsonIncludePaths.isArray())
```
Shouldn't we do something if `include-paths` is present but is not an array?

christianparpart replied:
At most, report (just like what if not having strings inside). Otherwise, still continue functioning. So: report & ignore (continue).

cameel replied:
Not sure if I missed it earlier, or if it changed, but in the current code `jsonIncludePaths.isArray() == false` does not result in an `Invalid JSON configuration passed.` report.

cameel replied:
Looking at it now, I'm still not sure if the change we did on the call is 100% correct. I.e. now it will add a trace when settings contains `"include-paths"` key but it's `null`. Wouldn't it be better to treat it the same way as the key not being there?

christianparpart replied:
@cameel isn't that subjective? I don't mind, and I have no too strong opinion about it. Feels minor to me, I prefer to be strict.  And if someone explicitly sets it to null, to be honest, to me that reads like he doesn't want any include paths. So the code IMHO(!) semantically does make sense, doesn't it?

cameel replied:
I'm fine with being strict, just not sure if the LSP clients are actually that strict themselves. And if they aren't, we'll have to adapt to them not they to us. I can see them just not being consistent and deciding to set a field to `null` instead of removing it in some cases. We had been doing that ourselves in JSON exports for a long time.

Anyway, I also think this is minor. Just wanted to let you know that this could potentially lead to a false-positive warning, depending on what clients do.

cameel commented on libsolidity/lsp/LanguageServer.cpp:
```diff
@@ -101,14 +101,29 @@ Json::Value LanguageServer::toJson(SourceLocation const& _location)
+	Json::Value jsonIncludePaths = _settings["include-paths"];
+	if (jsonIncludePaths && jsonIncludePaths.isArray())
+	{
+		std::vector<boost::filesystem::path> includePaths;
```
```suggestion
		vector<boost::filesystem::path> includePaths;
```

cameel commented on libsolidity/lsp/LanguageServer.cpp:
```diff
@@ -101,14 +101,29 @@ Json::Value LanguageServer::toJson(SourceLocation const& _location)
+	if (jsonIncludePaths && jsonIncludePaths.isArray())
+	{
+		std::vector<boost::filesystem::path> includePaths;
+		for (Json::Value jsonPath: jsonIncludePaths)
```
Shouldn't this be a reference? Or does it actually dispense copies?
```suggestion
		for (Json::Value const& jsonPath: jsonIncludePaths)
```

cameel commented on libsolidity/lsp/LanguageServer.cpp:
```diff
@@ -101,14 +101,29 @@ Json::Value LanguageServer::toJson(SourceLocation const& _location)
+		std::vector<boost::filesystem::path> includePaths;
+		for (Json::Value jsonPath: jsonIncludePaths)
+		{
+			if (!jsonPath.isString())
+			{
+				m_client.trace("Invalid JSON configuration passed. \"include-paths\" must be an array of strings.");
+				continue;
+			}
+			includePaths.emplace_back(boost::filesystem::path(jsonPath.asString()));
```
Isn't this equivalent to a simple `if`/`else`?
```suggestion
			if (!jsonPath.isString())
				m_client.trace("Invalid JSON configuration passed. \"include-paths\" must be an array of strings.");
			else
				includePaths.emplace_back(boost::filesystem::path(jsonPath.asString()));
```

christianparpart replied:
I wasn't sure about what style to chose. But yes ;)

cameel commented on libsolidity/lsp/LanguageServer.cpp:
```diff
@@ -101,14 +101,29 @@ Json::Value LanguageServer::toJson(SourceLocation const& _location)
+		{
+			if (!jsonPath.isString())
+			{
+				m_client.trace("Invalid JSON configuration passed. \"include-paths\" must be an array of strings.");
```
If there are 10 non-string values, you'll get 10 identical messages. Wouldn't it be better to either issue just one (about the whole array) or have each one mention the invalid value?

Marenz commented on Changelog.md:
```diff
@@ -13,6 +13,7 @@ Compiler Features:
  * Commandline Interface: option ``--pretty-json`` works also with the following options: ``--abi``, ``--asm-json``, ``--ast-compact-json``, ``--devdoc``, ``--storage-layout``, ``--userdoc``.
  * SMTChecker: Support ``abi.encodeCall`` taking into account the called selector.
  * Language Server: Allow full filesystem access to language server.
+ * Language Server: Always adds ``{project_root}/node_modules`` to include search paths.
```
I suggest to use `Always add` instead of `Always adds` as it is more consistent with the other entries

Marenz commented on libsolidity/lsp/LanguageServer.cpp:
```diff
@@ -101,14 +101,29 @@ Json::Value LanguageServer::toJson(SourceLocation const& _location)
+		{
+			if (!jsonPath.isString())
+			{
+				m_client.trace("Invalid JSON configuration passed. \"include-paths\" must be an array of strings.");
```
Should this be maybe a proper error with throwing and all?

christianparpart replied:
Not with throwing, at most reporting (trace does report towards the client), as this is an error caused by the client.

Marenz replied:
Then maybe reporting it at a higher log level tahn trace?

Marenz reviewed (dismissed):
In the commit with the descrp. `[lsp] Fixes sourceUnitNameToUri() path path resolving in relation to ... …include paths being used.`
Maybe one `path` is enough in the commit message? :P

cameel commented on libsolidity/lsp/FileRepository.cpp:
```diff
@@ -34,9 +34,25 @@ using namespace solidity::frontend;
+	// Prefix relative paths with base-path.
+	for (auto& path: _paths)
+		if (path.is_relative())
+			path = m_basePath / path;
```
Not sure if this makes sense. Normally the include paths are relative to working dir, not base path.

So this kinda depends on how LSP handles working dir. If it ignores it completely, this might be ok but if it's affected by it in anyway then probably not.

christianparpart replied:
the `m_basePath` does contain the project's root directory, as the information is passed to the server from the client (IDE).

cameel replied:
But does the working dir of the LSP have any influence on its path resolution?

cameel commented on libsolidity/lsp/FileRepository.cpp:
```diff
@@ -34,9 +34,25 @@ using namespace solidity::frontend;
+	setIncludePaths(move(_includePaths));
+}
+
+void FileRepository::setIncludePaths(std::vector<boost::filesystem::path> _paths)
```
This is not a simple setter. It makes some opinionated decisions. I think we should rename it to make it very clear:.
```suggestion
void FileRepository::setUpIncludePaths(std::vector<boost::filesystem::path> _requestedPaths)
```

cameel replied:
What about this one? Looks like it got lost among other comments.

cameel commented on libsolidity/lsp/FileRepository.cpp:
```diff
@@ -51,6 +67,8 @@ string FileRepository::sourceUnitNameToUri(string const& _sourceUnitName) const
 		return "file:///" + _sourceUnitName;
 	else if (_sourceUnitName.find("/") == 0)
 		return "file://" + _sourceUnitName;
+	else if (auto const resolvedPath = tryResolvePath(_sourceUnitName); resolvedPath.message().empty())
```
The actual condition in such ifs is really easy to miss. I'd split that in two lines (or just avoid this syntax):
```suggestion
	else if (
		auto const resolvedPath = tryResolvePath(_sourceUnitName);
		resolvedPath.message().empty()
	)
```

christianparpart replied:
We have much longer lines than that. I don't think that's particularly unreadable. But this is subjective. I'm not sure.

cameel replied:
For me it's not about the line length but about the condition blending in with the variable declaration. In most cases in our code such `if`s do not have the condition part after `;` and the condition is the variable. If it's not then IMO it's very easy to miss (I caught myself on that many times) and I think the condition should always be on a line of its own to make it more prominent.

cameel commented on libsolidity/lsp/FileRepository.cpp:
```diff
@@ -51,6 +67,8 @@ string FileRepository::sourceUnitNameToUri(string const& _sourceUnitName) const
 		return "file:///" + _sourceUnitName;
 	else if (_sourceUnitName.find("/") == 0)
 		return "file://" + _sourceUnitName;
+	else if (auto const resolvedPath = tryResolvePath(_sourceUnitName); resolvedPath.message().empty())
```
Isn't this supposed to kick in when the resolved path is **not** empty?
```suggestion
	else if (auto const resolvedPath = tryResolvePath(_sourceUnitName); !resolvedPath.message().empty())
```

cameel commented on libsolidity/lsp/FileRepository.cpp:
```diff
@@ -51,6 +67,8 @@ string FileRepository::sourceUnitNameToUri(string const& _sourceUnitName) const
+	else if (auto const resolvedPath = tryResolvePath(_sourceUnitName); resolvedPath.message().empty())
+		return "file://" + resolvedPath.get().generic_string();
 	else
 		return "file://" + m_basePath.generic_string() + "/" + _sourceUnitName;
```
Does `generic_string()` always strip the trailing `/`? What happens when base path is `/`? Won't the resulting path be interpreted as UNC path giving weird results?

christianparpart replied:
I just tried that locally in my browser and having too many leading slashes in your file-URI works. FF simply ignores, Chrome does rewrite the URL.

cameel commented on libsolidity/lsp/FileRepository.cpp:
```diff
@@ -51,6 +67,8 @@ string FileRepository::sourceUnitNameToUri(string const& _sourceUnitName) const
 		return "file:///" + _sourceUnitName;
```
Regarding the first part of the `if` above (I can't comment on it directly):
```c++
if (m_sourceUnitNamesToUri.count(_sourceUnitName))
		return m_sourceUnitNamesToUri.at(_sourceUnitName);
```
Is the assumption that this is already an URI, i.e. starts with `protocol://`? If so then maybe we should have an assert here?

christianparpart replied:
Yes, right below, the function named `setSourceByUri`  This is called in a context where the remote end is providing us URIs. I'm not sure an assert() is really necessary (?) as in that case, the client would just receive as URI what it initially has sent as URI (if it could be mapped to a local file). Hmmm.

cameel replied:
The thing I'm not sure about is: can a non-URI end up in `m_sourceUnitNamesToUri`? My impression from this function is that it's trying hard to always have the result prefixed with `file://` but it wasn't obvious to me if this particular `if` always guarantees that prefix too. My guess is yes and if that's the case, an assert would make it obvious. If not, I think it needs a comment saying how/why a non-URI can end up in `m_sourceUnitNamesToUri`.

christianparpart replied:
done

cameel commented on libsolidity/lsp/FileRepository.cpp:
```diff
@@ -51,6 +67,8 @@ string FileRepository::sourceUnitNameToUri(string const& _sourceUnitName) const
 		return "file:///" + _sourceUnitName;
 	else if (_sourceUnitName.find("/") == 0)
 		return "file://" + _sourceUnitName;
```
Doesn't this break resolution for files that are placed directly under one of the include dirs? I.e. if you have LSP set up equivalent to `--base-path /a/ --include-path b/c` then you won't be able to do `import "x.sol"` to import `/a/b/c/x.sol`, it will look for `/a/x.sol` instead.

christianparpart replied:
@cameel: If at all, I think this if-stmt should be removed all together. Meaning, it should not happen (AFAICT). I might be wrong. But if at all, I'd prefer to remove this in a seperate PR, because these lines are unrelated to the actual PR that should not become too big. What do you think?

cameel replied:
Fair enough. If it was there already I think it's fine to address it separately.

christianparpart replied:
@cameel I've added a test (bad-include) and it triggers a diagnostic that the given file "test.sol" could not be found.

cameel replied:
ok, so you'll be fixing that in a separate PR? I think it would make sense to report it as an issue unless you'll be doing that right away.

cameel replied:
I see this has been removed so marking as resolved.

cameel commented on test/libsolidity/lsp/include-paths/default-include.sol:
```diff
@@ -0,0 +1,12 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity >=0.8.0;
+
+import "my-module/test.sol";
```
Can you add a test that imports just `test.sol`?

christianparpart replied:
We have general import tests already for import path names that are meant to be relatively imported. What do you think should we expect of your import-test.sol-testcase?

cameel replied:
I expect it to trigger the problem I mentioned in https://github.com/ethereum/solidity/pull/12994#discussion_r874118206.

cameel commented:
> I think we could auto-detect at least the well known frameworks and adjust the include paths accordingly (on top of manually added ones, if added). What do you think, @cameel ?

I'm not really convinced that this should be baked into the compiler. It will lead to weird behavior in corner cases, when you're not on the "happy path" with a well supported framework and old compiler binaries will be less usable with new frameworks. I think we should instead give frameworks a way to provide this info to the compiler. I think that some non-framework-specific config file that the LSP expects to find under the base path might be a better idea.

chriseth commented on libsolidity/lsp/FileRepository.cpp:
```diff
@@ -34,9 +34,25 @@ using namespace solidity::frontend;
+			path = m_basePath / path;
+
+	// Always have {projectRoot}/node_modules in the include-paths.
+	_paths.emplace_back(m_basePath / "node_modules");
```
I don't think this is a good idea. As discussed in the meeting, I think it would be better to have `{"PROJECT_ROOT/node_modules"}` as the set of default includes paths. If you do it as it is done now there is no way to remove `node_modules` from the include paths.

christianparpart replied:
done.

cameel commented:
`t_win_soltest` is failing. And I just noticed that it runs LSP tests, which I didn't expect. I think this was originally a separate step and it made more sense that way. Then you can tell just from the job name what failed - I wouldn't even bother looking inside if I saw it's a failing LSP test in a LSP-related PR. And with this running after soltest, you have to first wait 20+ minutes for the Windows binary, and then 15+ minutes for soltest to finish. You'd get feedback much faster if this ran independently.

cameel commented on Changelog.md:
```diff
@@ -8,6 +8,8 @@ Language Features:
  * LSP: Add rudimentary support for semantic highlighting.
  * Yul Optimizer: Improve inlining heuristics for via IR code generation and pure Yul compilation.
+ * Language Server: Always add ``{project_root}/node_modules`` to include search paths.
+ * Language Server: Adds support for configuring extra include paths ``include-paths`` JSON settings object that can be passed during LSP configuration stage.
```
```suggestion
 * Language Server: Adds support for configuring ``include-paths`` JSON settings object that can be passed during LSP configuration stage.
```

cameel commented on test/lsp.py:
```diff
@@ -562,7 +562,11 @@ def test_diagnostics(self):
                 testname, sub_dir = split_path(testname_and_subdir)
 
-                expected_diagnostics = tests[testname]
+                # Only add expected diagnostics, if available.
+                expected_diagnostics = {}
+                if testname in tests:
+                    expected_diagnostics = tests[testname]
```
```suggestion
                expected_diagnostics = tests.get(testname, {})
```

cameel commented on test/libsolidity/lsp/include-paths/default-include.sol:
```diff
@@ -0,0 +1,14 @@
+    }
+}
+// ----
+// file:///home/christianparpart/work/solidity/test/libsolidity/lsp/node_modules/my-module/test:
```
This path looks like it's specific to your system. It shouldn't be here, should it?

christianparpart replied:
I was addressing this to @Marenz already. He can improve `lsp.py` seperately, I'd say. :)

cameel replied:
But is this something enforced by the test suite? If so, won't this fail for anyone but you? If not, why not remove it?

christianparpart replied:
@Marenz what's that line exactly there for? Maybe you can jump in?

Marenz replied:
I am not sure how or about what you two are misunderstanding each other, but yes, this path should be relative to the root dir or rather the test dir. I believe chris pushed this only so I can investigate.

cameel commented on libsolidity/lsp/LanguageServer.cpp:
```diff
@@ -146,14 +146,30 @@ Json::Value LanguageServer::toJson(SourceLocation const& _location)
+			else
+				typeFailures++;
+		}
+		if (typeFailures)
```
I think this would be much more readable. From the name it looks like a boolean but it isn't.
```suggestion
		if (typeFailureCount > 0)
```

cameel commented on libsolidity/lsp/FileRepository.cpp:
```diff
@@ -69,6 +91,52 @@ void FileRepository::setSourceByUri(string const& _uri, string _source)
+	vector<boost::filesystem::path> candidates;
+	vector<reference_wrapper<boost::filesystem::path const>> prefixes = {m_basePath};
+	prefixes += (m_includePaths | ranges::to<vector<reference_wrapper<boost::filesystem::path const>>>);
+	auto const defaultInclude = m_basePath / "node_modules";
```
This looks like a better place for adding special paths than `sourceUnitNameToUri()` but this still makes it impossible not to have any include paths at all - and this is a very common case. Maybe a more logical place to add this would be `changeConfiguration()`? That function can discern between "include paths not set" (i.e. `null`) and "include paths set is empty" (i.e. `[]`).

Unless the LSP clients do not care and would just always send `[]`... What do you think?

christianparpart replied:
I think the current behavior is fine. I've tested it with both, and the default configuration should suffice (works like solidity-vscode extension by default works, too), but gives the ability to customize and also _not_ to include the above mentioned default include of "${projectRoot}/node_modules".

cameel replied:
But that's only if you include some other path, right? Because the condition below is `m_includePaths.empty()` and that's going to always add it if you have no include paths.

christianparpart replied:
Yes, it's going to add that default include only if the user hasn't supplied any. Do you explicitly want to support dropping the default include and have zero other include paths set? I think this case is rather artificial. The default include path is for ease-of-use, as that's most likely the most often used include path being used in the wild (AFAICT). If people chose to roll their own, they can do, which would not auto-add the default include path. Hmmm....

cameel replied:
Yeah, I think it would be best if one could drop the default path from the set. I mean, it probably won't hurt anyone in a vast majority of cases but it just does not seem quite right. If the client has a way to say explicitly "no extra paths" (and this depends on client's UI), the server should have a way to respect that. In some corner cases you could have a project that does not use `node_modules` for contracts but still have a stray dir there because you were installing some extra tools via npm.

Now that I think of it - is there a mechanism for the server to provide the "default" settings for the client to use as a base value? Because the "right" way to solve this IMO would be to just put `node_modules` path in these defaults so that user does not have to do anything but is also able to modify them.

Overall this is not a deal-breaker and I'm going to accept the PR regardless of this but I'm just thinking that maybe we could solve this better.

cameel replied:
Marking as resolved.

We discussed it on the call and looks like there is not way for the client to query the configuration. So my idea is not viable after all.

cameel commented on test/libsolidity/lsp/include-paths/bad_include.sol:
```diff
@@ -0,0 +1,11 @@
+pragma solidity >=0.8.0;
+
+   import "test.sol";
+// ^^^^^^^^^^^^^^^^^^ @BadInclude
```
As discussed on the chat, the include is valid, it's just that the file is missing - so the error (and test name) should say just that.

cameel replied:
Also, we should have a positive test for this case too. I.e. a file placed directly under base path or one of the include paths should be successfully imported.

christianparpart committed:
lsp: Adding test for custom include paths.

cameel commented on test/libsolidity/lsp/other-include-dir/otherlib/otherlib.sol:
```diff
@@ -0,0 +1,6 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity >=0.8.0;
+
+library MyRootLib
```
```suggestion
library OtherLib
```

cameel reviewed (commented):
State of the review after our call and subsequent changes:
1. There are two minor comments to resolve that I added now (lib name and trace for `null` settings).
2. Some tests are failing.
3. We agreed that there should be some tests covering the function for setting configuration and non-default include paths.

cameel commented on test/libsolidity/lsp/include-paths/using-custom-includes.sol:
```diff
@@ -0,0 +1,11 @@
+pragma solidity >=0.8.0;
+
+   import "otherlib/otherlib.sol";
+// ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ @SomeTag
```
Shouldn't this succeed? And what kind of error is `SomeTag`?

christianparpart replied:
Not in this test. But in the lsp.py there is an explicit test with the custom include path set, THERE it should and will pass. :)

cameel commented on test/lsp.py:
```diff
@@ -1381,6 +1381,37 @@ def get_test_tags(self, test_name: TestName, sub_dir=None, verbose=False):
+                'settings':
+                {
+                    'include-paths':
+                    [
```
```suggestion
                'settings': {
                    'include-paths': [
```

narghev commented:
Hi, just had this problem on nvim with newly installed version of `solc` using `brew`. 
```Version: 0.8.15+commit.e14f2714.Darwin.appleclang```

Wanted to check if this is released already or not.

cameel commented:
Yes, this will go into 0.8.16, which we are releasing tomorrow.
